README.md
# SMS Bulk Sender Reminder Service

Via an existing spreadsheet driven business process, automate the sending of SMS
reminders without impacting on existing processes (don't change the template),
but do same the business time by automating the sending of SMS reminders.

# Install


```
cp .env.example .env
```

```
python3 -m venv venv
. venv/bin/activate
pip install -r requirements.txt
```

# Run

```
. venv/bin/activate
cd src/web
flask --debug run
```


CHANGELOG.md
# v0.0.15 (Wed May 03 2023)

#### 🐛 Bug Fix

- #7 removed unused pandas module [#9](https://github.com/KarmaComputing/sms-bulk-sending/pull/9) ([@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 1

- [@chrisjsimpson](https://github.com/chrisjsimpson)

---

# v0.0.14 (Wed May 03 2023)

#### 🐛 Bug Fix

- Fix #7 remove duplicated requirements.txt [#8](https://github.com/KarmaComputing/sms-bulk-sending/pull/8) ([@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 1

- [@chrisjsimpson](https://github.com/chrisjsimpson)

---

# v0.0.13 (Wed May 03 2023)

#### 🐛 Bug Fix

- Fix #5 add confirmation step before sending SMS [#6](https://github.com/KarmaComputing/sms-bulk-sending/pull/6) ([@chrisjsimpson](https://github.com/chrisjsimpson))
- wip #5 tidy sms sending errors [#6](https://github.com/KarmaComputing/sms-bulk-sending/pull/6) ([@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 1

- [@chrisjsimpson](https://github.com/chrisjsimpson)

---

# v0.0.12 (Wed May 03 2023)

#### ⚠️ Pushed to `main`

- #3 fix use issue templates flash messages ([@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 1

- [@chrisjsimpson](https://github.com/chrisjsimpson)

---

# v0.0.11 (Wed May 03 2023)

:tada: This release contains work from a new contributor! :tada:

Thank you, null[@joeltejeda](https://github.com/joeltejeda), for all your work!

#### 🐛 Bug Fix

- 3 use template inheritance [#4](https://github.com/KarmaComputing/sms-bulk-sending/pull/4) ([@joeltejeda](https://github.com/joeltejeda) [@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 2

- [@chrisjsimpson](https://github.com/chrisjsimpson)
- [@joeltejeda](https://github.com/joeltejeda)

---

# v0.0.10 (Wed May 03 2023)

:tada: This release contains work from a new contributor! :tada:

Thank you, null[@joeltejeda](https://github.com/joeltejeda), for all your work!

#### 🐛 Bug Fix

- wip #1 adding env.example and printing the test info [#2](https://github.com/KarmaComputing/sms-bulk-sending/pull/2) ([@joeltejeda](https://github.com/joeltejeda) [@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 2

- [@chrisjsimpson](https://github.com/chrisjsimpson)
- [@joeltejeda](https://github.com/joeltejeda)

---

# v0.0.9 (Sat Apr 22 2023)

#### ⚠️ Pushed to `main`

- Update base.html ([@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 1

- [@chrisjsimpson](https://github.com/chrisjsimpson)

---

# v0.0.8 (Sat Apr 22 2023)

#### ⚠️ Pushed to `main`

- Added deploy.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added release.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added deploy.sh file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added .autorc file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added amber.yaml secrets file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added flask quickstart ([@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 1

- [@chrisjsimpson](https://github.com/chrisjsimpson)

---

# v0.0.7 (Sat Apr 22 2023)

#### ⚠️ Pushed to `main`

- Added deploy.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added release.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added deploy.sh file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added .autorc file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added amber.yaml secrets file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added flask quickstart ([@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 1

- [@chrisjsimpson](https://github.com/chrisjsimpson)

---

# v0.0.6 (Sat Apr 22 2023)

#### ⚠️ Pushed to `main`

- Added deploy.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added release.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added deploy.sh file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added .autorc file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added amber.yaml secrets file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added flask quickstart ([@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 1

- [@chrisjsimpson](https://github.com/chrisjsimpson)

---

# v0.0.5 (Sat Apr 22 2023)

#### ⚠️ Pushed to `main`

- Added deploy.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added release.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added deploy.sh file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added .autorc file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added amber.yaml secrets file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added flask quickstart ([@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 1

- [@chrisjsimpson](https://github.com/chrisjsimpson)

---

# v0.0.4 (Sat Apr 22 2023)

#### ⚠️ Pushed to `main`

- Added deploy.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added release.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added deploy.sh file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added .autorc file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added amber.yaml secrets file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added flask quickstart ([@chrisjsimpson](https://github.com/chrisjsimpson))

#### Authors: 1

- [@chrisjsimpson](https://github.com/chrisjsimpson)

---

# v0.0.3 (Sat Apr 22 2023)

#### ⚠️ Pushed to `main`

- Added deploy.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added release.yml file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added deploy.sh file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added .autorc file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added amber.yaml secrets file ([@chrisjsimpson](https://github.com/chrisjsimpson))
- Added flask quickstart ([@chrisjsimpson](https://github.com/chrisjsimpson))


extract.py
from openpyxl import load_workbook
import os
from dotenv import load_dotenv
import requests

load_dotenv()  # take environment variables from .env.

FILENAME = os.getenv("FILENAME", None)
api_key = os.getenv("BUDGETSMS_API_KEY")
user_id = os.getenv("BUDGETSMS_USER_ID")
sender_id = os.getenv("BUDGETSMS_SENDER_ID")
username = os.getenv("BUDGETSMS_USERNAME")
UPLOAD_FOLDER = os.getenv("UPLOAD_FOLDER")

wb = load_workbook(filename=f"{UPLOAD_FOLDER}/{FILENAME}")

rows = list(wb._sheets[0].rows)

treatment_jobsheet_date = wb._sheets[0]["B1"].value

# Cut off row one with data, and row two with column headers
rows = rows[2:-1]

customer_ref_column_number = 2  # (Column C)

customers_with_upcoming_jobs = []

for row in rows:
    if row[customer_ref_column_number].value is None:
        print(
            "customer ref was None in Treatment Job Sheet, so stopping processing"  # noqa: E501
        )  # noqa: E501
        break
    customer_ref = row[customer_ref_column_number].value
    print(f"customer_ref {customer_ref}")
    customers_with_upcoming_jobs.append(customer_ref)


# Locate customer PhoneNumber based on customer_ref number inside
# customers_with_upcoming_jobs, get their phone number from sheet
# called 'Contacts_Prices_Comments' column 'PhoneNumber'
# (column 6 / column 'G')
# TODO we're **assumming** 'PhoneNumber' is a mobile, which is obviously
# wrong.

sheet_customer_list_contacts_prices_comments = list(wb._sheets[4].rows)
PhoneNumberColumn = 6

sheet_customer_list_contacts_prices_comments[1:-1]  # cut off heading

# Find the contact number for a given customer
numbers_to_contact = []

# Loop customers with upcoming jobs
for customer_ref in customers_with_upcoming_jobs:

    for row in sheet_customer_list_contacts_prices_comments:
        if customer_ref == row[0].value:
            # Locate customers phone number
            numbers_to_contact.append(row[PhoneNumberColumn].value)
message = "Hello, this is a test message from BudgetSMS."
# Send the message to each mobile number
for number in numbers_to_contact:
    url = "https://api.budgetsms.net/sendsms/"
    params = {
        "username": username,
        "userid": user_id,
        "handle": api_key,
        "from": sender_id,
        "to": number,
        "msg": message,
    }
    response = requests.get(url, params=params)
    if response.status_code == 200 and response.text != "ERR 1001":
        print(f"Message successfully sent to {number}.")
    else:
        print(
            f"Failed to send message to {number}. Response code: {response.status_code} Response text: {response.text}."
        )


src/web/utils.py
import os
from openpyxl import load_workbook
import requests


def extract_customer_numbers_from_spreadsheet(filename: str) -> list:
    """Extract customers numbers from spreadsheet
    retuns: list of numbers numbers_to_contact"""

    UPLOAD_FOLDER = os.getenv("UPLOAD_FOLDER")

    wb = load_workbook(filename=f"{UPLOAD_FOLDER}/{filename}")

    rows = list(wb._sheets[0].rows)

    # Cut off row one with data, and row two with column headers
    rows = rows[2:-1]

    customer_ref_column_number = 2  # (Column C)

    customers_with_upcoming_jobs = []

    for row in rows:
        if row[customer_ref_column_number].value is None:
            print(
                "customer ref was None in Treatment Job Sheet, so stopping processing"  # noqa: E501
            )  # noqa: E501
            break
        customer_ref = row[customer_ref_column_number].value
        print(f"customer_ref {customer_ref}")
        customers_with_upcoming_jobs.append(customer_ref)

    # Locate customer PhoneNumber based on customer_ref number inside
    # customers_with_upcoming_jobs, get their phone number from sheet
    # called 'Contacts_Prices_Comments' column 'PhoneNumber'
    # (column 6 / column 'G')
    # TODO we're **assumming** 'PhoneNumber' is a mobile, which is obviously
    # wrong.

    sheet_customer_list_contacts_prices_comments = list(wb._sheets[4].rows)
    PhoneNumberColumn = 6

    sheet_customer_list_contacts_prices_comments[1:-1]  # cut off heading

    # Find the contact number for a given customer
    numbers_to_contact = []

    # Loop customers with upcoming jobs
    for customer_ref in customers_with_upcoming_jobs:
        for row in sheet_customer_list_contacts_prices_comments:
            if customer_ref == row[0].value:
                # Locate customers phone number
                numbers_to_contact.append(row[PhoneNumberColumn].value)

    return numbers_to_contact


def send_sms(to_number: str, msg: str):
    print("About to send sms")
    # Send the message to each mobile number
    url = "https://api.budgetsms.net/sendsms/"
    BUDGETSMS_API_KEY = os.getenv("BUDGETSMS_API_KEY")
    BUDGETSMS_USER_ID = os.getenv("BUDGETSMS_USER_ID")
    BUDGETSMS_SENDER_ID = os.getenv("BUDGETSMS_SENDER_ID")
    BUDGETSMS_USERNAME = os.getenv("BUDGETSMS_USERNAME")

    params = {
        "username": BUDGETSMS_USERNAME,
        "userid": BUDGETSMS_USER_ID,
        "handle": BUDGETSMS_API_KEY,
        "from": BUDGETSMS_SENDER_ID,
        "to": to_number,
        "msg": msg,
    }
    response = requests.get(url, params=params)
    if response.status_code == 200 and "ERR" not in response.text:
        print(f"Message successfully sent to {to_number}.")
        return 200
    else:
        print(
            f"Failed to send message to {to_number}. Response code: {response.status_code} Response text: {response.text}."  # noqa: E501
        )
        return (
            500,
            f"Sorry the message couldn't be sent to the phone number: {to_number}",
        )


def allowed_file(filename):
    ALLOWED_EXTENSIONS = os.getenv("ALLOWED_EXTENSIONS")
    return "." in filename and filename.rsplit(".", 1)[1].lower() in ALLOWED_EXTENSIONS


src/web/app.py
from db import get_db, close_db
import sqlalchemy
from logger import log
from flask import (
    Flask,
    request,
    render_template,
    redirect,
    url_for,
    flash,
    session,
)  # noqa: E501
import os
from dotenv import load_dotenv
from utils import (
    send_sms,
    extract_customer_numbers_from_spreadsheet,
    allowed_file,
)
from werkzeug.utils import secure_filename


# Load app settings
load_dotenv()

app = Flask(__name__)
UPLOAD_FOLDER = os.getenv("UPLOAD_FOLDER", None)
app.config["SECRET_KEY"] = os.getenv("SECRET_KEY")
app.teardown_appcontext(close_db)


@app.route("/confirm")
def confirm():
    """Confirm to the user that they are happy for the SMS messages to
    be sent"""
    try:
        if "confirm" in request.args:
            for number in session["numbers_to_contact"]:
                message_sent = send_sms(number, "Mow your lawn")
                if message_sent[0] == 500:
                    flash(message_sent[1])
                    flash("Tip: No spaces permitted in number")
                    flash(
                        "Do not use the + before the international countrycode."  # noqa: E501
                    )  # noqa: E501
                    flash(
                        "Omit the first 0 from the number. Example: 31612345678"  # noqa: E501
                    )  # noqa: E501

    except Exception() as e:
        log.error(f"Error sending SMS messages: {e}")
        return "Error sending SMS messages", 500
    return render_template("confirmation.html")


@app.route("/health")
def health():
    """check health of app by
    - checking for valid (working) database connection
    - The 200 http response can also be used as a health indicator
    """
    log.info("Checking /health")
    db = get_db()
    health = "BAD"
    try:
        result = db.execute("SELECT NOW()")
        result = result.one()
        health = "OK"
        log.info(
            f"/health reported OK including database connection: {result}"
        )  # noqa: E501
    except sqlalchemy.exc.OperationalError as e:
        msg = f"sqlalchemy.exc.OperationalError: {e}"
        log.error(msg)
    except Exception as e:
        msg = f"Error performing healthcheck: {e}"
        log.error(msg)

    return health


@app.route("/", methods=["GET", "POST"])
def index():
    if request.method == "POST":
        file = request.files["file"]
        if "file" not in request.files or file.filename == "":
            flash("No file selected")
            return render_template("index.html")
        elif allowed_file(file.filename):
            # Save the uploaded file to the uploads folder
            filename = secure_filename(file.filename)
            file.save(os.path.join(UPLOAD_FOLDER, filename))
            numbers_to_contact = extract_customer_numbers_from_spreadsheet(
                filename
            )  # noqa: E501

            # Store numbers to contact in session
            session["numbers_to_contact"] = numbers_to_contact

            # Redirect to confirmation before sending sms messages
            return redirect(url_for("confirm"))
        else:
            flash("The file extension is not allowed")
            return render_template("index.html")

    return render_template("index.html")


@app.route("/download/<filename>")
def download(filename=None):
    if filename is None:
        return render_template("upload.html")
    return render_template("finished.html")
